README
NAME
    Object::Previous - find the instance of the object that called your
    function

SYNOPSIS
        package Human;
        use Object::Previous;

        sub new { bless {hit_points=>(7+int rand 6)} }
        sub hurt_us {
            my $body = shift;
               $body->{hit_points} -= shift;

            if( (int rand 10) == 0 ) {
                # every once in a while, damaging bodies hurts the sword:

                my $sword = previous_object();
                   $sword->hurt_us(1+int rand 4);
            }
        }

        package Sword;
        sub new { bless {hit_points=>2} }
        sub hurt_human {
            my $sword = shift;
            my $target = shift;

            $target->hurt_us( 1+int rand 8 );
        }
        sub hurt_us {
            my $sword = shift;
               $sword->{hit_points} -= shift;

            if( $sword->{hit_points} <= 0 ) {
                warn "the attacker's sword broke!";
            }
        }

previous_object()
    previous_object() either returns the blessed ref of the caller or undef
    if it is not possible to find it.

CAVEATS
    If you tinker with the @_ in the caller object, previous_object() won't
    work. Curiously, certain ways of tinkinkering don't hurt and others do.

        my $self = shift; # doesn't hurt previous_object()
        shift while @_;   # doesn't hurt previous_object()
        splice @_, 0, 30; # doesn't hurt previous_object()

        unshift @_, "borked"; # breaks previous_object();
        @_ = ();              # breaks previous_object();

AUTHOR(S)
    Most of the code was ripped from Perl and from perlmonks.org, but it was
    glued together by me.

    Paul Miller <paul@cpan.org>

    I am using this software in my own projects... If you find bugs, please
    please please let me know. :) Actually, let me know if you find it handy
    at all. Half the fun of releasing this stuff is knowing that people use
    it.

COPYRIGHT
    Copyright (c) 2007 Paul Miller

    Licensed under the same terms as Perl itself.

SEE ALSO
    perl(1), Devel::Stacktrace, perlmonks.org, cop.h, pp_ctl.c



Changes
1.1011: Wed Jul 08 2009
   - worst bug ever: I'm jettero@cpan.org, not paul@cpan.org

1.1010: Wed Jul 08 2009
   - Doc bugs (very indirectly) found by Hans Dieter Pearcey

1.2: Sun May  3 18:33:49 EDT 2009
   - doc updates, dist updates, etc

1.1: Wed Oct 29 08:33:39 EDT 2008
   - roflmao: I quit using version.pm but left it in the pm

1.1: Thu Aug  7 15:49:49 EDT 2008
    - version.pm is incompatible with EU::MM and
    - removed the Autoloader from the .pm

1.0.0:  Sat Jun  7 19:59:19 EDT 2008
    - finished the pure perl and XS versions


Previous.pm
package Object::Previous;

use strict;
use warnings;
use Carp;

require Exporter;
use base 'Exporter';

our $VERSION = "1.1011";
our @EXPORT = qw(previous_object); ## no critic

sub previous_object {};
sub import {
    if( @_==1 or $_[1] !~ m/(:?pure|perl|pl)/ ) {
        eval {
            require XSLoader;
            XSLoader::load('Object::Previous', $VERSION);
        };

        if( $@ ) {
            warn "couldn't load _xs version: $@";
            *previous_object = *previous_object_pl;

        } else {
            *previous_object = *previous_object_xs;
        }

    } else {
        splice @_, 1, 1;
        *previous_object = *previous_object_pl;
    }

    goto &Exporter::import;
}

sub previous_object_pl {
    my @foo = do { package DB; @DB::args=(); caller(2) }; ## no critic

    # NOTE: this doesn't work if, in that previous object, you were to do this:
    #
    #   unshift @_, "borked".
    #
    # The result is that you'd get "borked" instead of the blessed ref of the caller object

    # NOTE: I call this pure-perl vesion The Chromatic Way, but it's really the Devel::StackTrace way see:
    #  - http://perlmonks.org/?node_id=690713
    #  - http://perlmonks.org/?node_id=690795

    return $DB::args[0];
}

1;


pp_caller.c

PP(pp_caller) {
    dSP;
    register I32 cxix = dopoptosub(cxstack_ix);
    register const PERL_CONTEXT *cx;
    register const PERL_CONTEXT *ccstack = cxstack;
    const PERL_SI *top_si = PL_curstackinfo;
    I32 gimme;
    const char *stashname;
    I32 count = 0;

    if (MAXARG)
        count = POPi;

    for (;;) {
        /* we may be in a higher stacklevel, so dig down deeper */
        while (cxix < 0 && top_si->si_type != PERLSI_MAIN) {
            top_si = top_si->si_prev;
            ccstack = top_si->si_cxstack;
            cxix = dopoptosub_at(ccstack, top_si->si_cxix);
        }

        if (cxix < 0) {
            if (GIMME != G_ARRAY) {
                EXTEND(SP, 1);
                RETPUSHUNDEF;
            }

            RETURN;
        }

        /* caller() should not report the automatic calls to &DB::sub */
        if (PL_DBsub && GvCV(PL_DBsub) && cxix >= 0 && ccstack[cxix].blk_sub.cv == GvCV(PL_DBsub))
            count++;

        if (!count--)
            break;

        cxix = dopoptosub_at(ccstack, cxix - 1);
    }

    cx = &ccstack[cxix];
    if (CxTYPE(cx) == CXt_SUB || CxTYPE(cx) == CXt_FORMAT) {
        const I32 dbcxix = dopoptosub_at(ccstack, cxix - 1);

        /* We expect that ccstack[dbcxix] is CXt_SUB, anyway, the
           field below is defined for any cx. */
        /* caller() should not report the automatic calls to &DB::sub */

        if (PL_DBsub && GvCV(PL_DBsub) && dbcxix >= 0 && ccstack[dbcxix].blk_sub.cv == GvCV(PL_DBsub))
            cx = &ccstack[dbcxix];
    }

    stashname = CopSTASHPV(cx->blk_oldcop);
    if (GIMME != G_ARRAY) {
        EXTEND(SP, 1);

        // PAUL: scalar context, seems to skip the stuff I want

        if (!stashname)
            PUSHs(&PL_sv_undef);

        else {
            dTARGET;
            sv_setpv(TARG, stashname);
            PUSHs(TARG);
        }

        RETURN;
    }

    EXTEND(SP, 10);

    // PAUL: my @foo = caller(2)
    // PAUL: stash == namespace apparently $foo[0]
    if (!stashname) PUSHs(&PL_sv_undef);
    else            PUSHs(sv_2mortal(newSVpv(stashname, 0)));

    PUSHs(sv_2mortal(newSVpv(OutCopFILE(cx->blk_oldcop), 0))); // the filename $foo[1]
    PUSHs(sv_2mortal(newSViv((I32)CopLINE(cx->blk_oldcop)))); // line $foo[2]

    if (!MAXARG)
        RETURN;

    if (CxTYPE(cx) == CXt_SUB || CxTYPE(cx) == CXt_FORMAT) {
        GV *cvgv = CvGV(ccstack[cxix].blk_sub.cv);

        /* So is ccstack[dbcxix]. */

        // PAUL: this is probably what we're looking for... $foo[3] is the name of the subroutine
        //       $foo[4] is "hasargs:" whether a new @_ was set up for the frame

        if (isGV(cvgv)) {
            SV * const sv = NEWSV(49, 0);
            gv_efullname3(sv, cvgv, Nullch);
            PUSHs(sv_2mortal(sv));
            PUSHs(sv_2mortal(newSViv((I32)cx->blk_sub.hasargs)));
        }

        else {
            PUSHs(sv_2mortal(newSVpvn("(unknown)",9)));
            PUSHs(sv_2mortal(newSViv((I32)cx->blk_sub.hasargs)));
        }

    } else {
        PUSHs(sv_2mortal(newSVpvn("(eval)",6)));
        PUSHs(sv_2mortal(newSViv(0)));
    }

    gimme = (I32)cx->blk_gimme;

    if (gimme == G_VOID) PUSHs(&PL_sv_undef);
    else                 PUSHs(sv_2mortal(newSViv(gimme & G_ARRAY)));

    if (CxTYPE(cx) == CXt_EVAL) {
        if (cx->blk_eval.old_op_type == OP_ENTEREVAL) {
            /* eval STRING */
            PUSHs(cx->blk_eval.cur_text);
            PUSHs(&PL_sv_no);

        } else if (cx->blk_eval.old_namesv) {
            /* require */
            PUSHs(sv_2mortal(newSVsv(cx->blk_eval.old_namesv)));
            PUSHs(&PL_sv_yes);

        } else {
            /* eval BLOCK (try blocks have old_namesv == 0) */
            PUSHs(&PL_sv_undef);
            PUSHs(&PL_sv_undef);
        }

    } else {
        PUSHs(&PL_sv_undef);
        PUSHs(&PL_sv_undef);
    }

    if (CxTYPE(cx) == CXt_SUB && cx->blk_sub.hasargs && CopSTASH_eq(PL_curcop, PL_debstash)) {
        AV * const ary = cx->blk_sub.argarray;
        const int off = AvARRAY(ary) - AvALLOC(ary);

        if (!PL_dbargs) {
            GV* tmpgv;
            PL_dbargs = GvAV(gv_AVadd(tmpgv = gv_fetchpv("DB::args", TRUE, SVt_PVAV)));
            GvMULTI_on(tmpgv);
            AvREAL_off(PL_dbargs);	/* XXX should be REIFY (see av.h) */
        }

        if (AvMAX(PL_dbargs) < AvFILLp(ary) + off)
            av_extend(PL_dbargs, AvFILLp(ary) + off);

        Copy(AvALLOC(ary), AvARRAY(PL_dbargs), AvFILLp(ary) + 1 + off, SV*);
        AvFILLp(PL_dbargs) = AvFILLp(ary) + off;
    }

    /* XXX only hints propagated via op_private are currently
     * visible (others are not easily accessible, since they
     * use the global PL_hints) */
    PUSHs(sv_2mortal(newSViv((I32)cx->blk_oldcop->op_private & HINT_PRIVATE_MASK)));

    // PAUL: you can drop an arbitrary scope in C?
    {
        SV * mask ;
        SV * old_warnings = cx->blk_oldcop->cop_warnings ;

        if  (old_warnings == pWARN_NONE || (old_warnings == pWARN_STD && (PL_dowarn & G_WARN_ON) == 0))
            mask = newSVpvn(WARN_NONEstring, WARNsize) ;

        else if (old_warnings == pWARN_ALL || (old_warnings == pWARN_STD && PL_dowarn & G_WARN_ON)) {
            /* Get the bit mask for $warnings::Bits{all}, because
             * it could have been extended by warnings::register */
            SV **bits_all;
            HV *bits = get_hv("warnings::Bits", FALSE);

            if (bits && (bits_all=hv_fetch(bits, "all", 3, FALSE))) {
                mask = newSVsv(*bits_all);

            } else {
                mask = newSVpvn(WARN_ALLstring, WARNsize) ;
            }

        } else mask = newSVsv(old_warnings);

        PUSHs(sv_2mortal(mask));
    }

    RETURN;
}


PreviousObject.pm
package Object::PreviousObject;

use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;
use base 'Exporter';
use version; our $VERSION = qv('1.0.0');

our @EXPORT = qw(previous_object);

sub previous_object {};
sub import {
    if( @_==1 or $_[1] !~ m/(:?pure|perl)/ ) {
        eval {
            require XSLoader;
            XSLoader::load('Object::PreviousObject', $VERSION);
        };

        if( $@ ) {
            warn "couldn't load _xs version: $@";
            *previous_object = *previous_object_perl;

        } else {
            *previous_object = *previous_object_xs;
        }

    } else {
        splice @_, 1, 1;
        *previous_object = *previous_object_perl;
    }

    goto &Exporter::import;
}

sub previous_object_perl {
    my @foo = do { package DB; @DB::args=(); caller(2) };

    # NOTE: this doesn't work if, in that previous object, you were to do this:
    #
    #   unshift @_, "borked".
    #
    # The result is that you'd get "borked" instead of the blessed ref of the caller object

    # NOTE: I call this pure-perl vesion The Chromatic Way, but it's really the Devel::StackTrace way see:
    #  - http://perlmonks.org/?node_id=690713
    #  - http://perlmonks.org/?node_id=690795

    $DB::args[0];
}

1;

__END__

=head1 NAME

Object::PreviousObject - find the instance of the object that called your function

=head1 SYNOPSIS

    package Human;
    use Object::PreviousObject;

    sub hurt_us {
        my $body = shift;
           $body->{hit_points} -= shift;

        if( (int rand 10) == 0 ) {
            # every once in a while, damaging bodies hurts the sword:

            my $sword = previous_object();
               $sword->hurt_us(1+int rand4);
        }
    }

    package Sword;
    sub hurt_human {
        my $sword = shift;
        my $target = shift;

        $target->hurt_us( 1+int rand 8 );
    }
    sub hurt_us {
        my $sword = shift;
           $sword->{hit_points} -= shift;

        if( $sword->{hit_points} <= 0 ) {
            warn "the attacker's sword broke!";
        }
    }

=head1 previous_object()

previous_object() either returns the blessed ref of the caller or
undef if it is not possible to find it.

=head1 CAVEATS

If you tinker with the @_ in the caller object, previous_object()
won't work.  Curiously, certain ways of tinkinkering don't hurt
and others do.

    my $self = shift; # doesn't hurt previous_object()
    shift while @_;   # doesn't hurt previous_object()
    splice @_, 0, 30; # doesn't hurt previous_object()

    unshift @_, "borked"; # breaks previous_object();
    @_ = ();              # breaks previous_object();

=head1 AUTHOR(S)

Most of the code was ripped from Perl and from perlmonks.org, but
it was glued together by me.

Paul Miller <paul@cpan.org>

I am using this software in my own projects...  If you find bugs,
please please please let me know. :) Actually, let me know if you
find it handy at all.  Half the fun of releasing this stuff is
knowing that people use it.

=head1 COPYRIGHT

Copyright (c) 2007 Paul Miller -- LGPL [attached] where possible,
however, since much of this code was ripped from Perl itself, it
probably has to follow the perl license?  My intention is to use
the LGPL if possible.

=head1 SEE ALSO

perl(1), Devel::Stacktrace, perlmonks.org

=cut
